reqwest-retry/src/lib.rs

```diff
@@ -30,10 +30,24 @@ mod retryable;
 mod retryable_strategy;
 
 pub use retry_policies::{policies, Jitter, RetryDecision, RetryPolicy};
+use thiserror::Error;
 
 pub use middleware::RetryTransientMiddleware;
 pub use retryable::Retryable;
 pub use retryable_strategy::{
     default_on_request_failure, default_on_request_success, DefaultRetryableStrategy,
     RetryableStrategy,
 };
+
+/// Custom error type to attach the number of retries to the error message.
+#[derive(Debug, Error)]
+pub enum RetryError {
+    #[error("Request failed after {retries} retries")]
+    WithRetries {
+        retries: u32,
+        #[source]
+        err: reqwest_middleware::Error,
+    },
+    #[error(transparent)]
+    Error(reqwest_middleware::Error),
```

Comment:
suggestion:

using an enum here is going to make it harder for a user to match on this error now, because they have to handle an extra branch that is still just a `reqwest_middleware::Error`

it seems completely unnecessary, why don't you just make this a struct that has a retry count and an error?
the only place where `RetryError::Error` was constructed, we actually do have a retry count that we could supply.

Reply:
for example, suppose I'm a user, and i want to do further error handling and call `is_body` on the underlying `Reqwest::Error`. https://docs.rs/reqwest/latest/reqwest/struct.Error.html#method.is_body
With this enum that requires a lot more boilerplate code

Reply:
We need the enum so we can insert the error message about the number of retries when we did them, while not showing an irrelevant "Failed after 0 retries" otherwise.

Reply:
that doesn't seem like such a big problem. maybe you could customize `Display` so that it doesn't show when the number is 0? the enum is going to create boilerplate at any call-site that wants to further inspect the error

Reply:
We can't skip the display on 0 because it `ToString` has to return a string even if it's an empty string, the only way i know of how to skip a step in the error trace is `error(transparent)`.

Reply:
> We can't skip the display on 0 because it ToString has to return a string even if it's an empty string

i think maybe you misunderstood me, you could do something like this

```
pub struct Error {
   err: reqwest::Error,
   retries: u32,
}
```

then you could implement `Display` like

```
impl Display for Error {
    fn fmt(&self, f: ...) -> ... {
       if self.retries == 0 {
          write!(f, "{}", self.err)
       } else {
          write!(f, "Failed after {} retries: {}", self.retries, self.err)
       }
    }
}
```

or similar